### PR TITLE
Management command to download Instructor Profile images from Drupal

### PR DESCRIPTION
**Pull request by staubina**
[DISCO-283]

This command will download all the images from the profile_image_url
field from Drupal into the Discovery data store and update the model.

### CODE REVIEW

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+IMAGE_TYPES = {
+    'image/jpeg': 'jpg',
+    'image/png': 'png',
+}
+
+
+class Command(BaseCommand):
+
+    def add_arguments(self, parser):
+        super().add_arguments(parser)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py]: You can just drop this method if all it does is call it's super.

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            profile_image_url='',
+        )
+
+        logger.info('There are {person_count} instructor images to download'.format(person_count=persons.count()))
+        failed_image_pulls = []
+        success_count = 0
+        for person in persons:
+            if person.profile_image and person.profile_image_url == person.profile_image.url:
+                logger.info('image already local not retrieving image for Instructor {instructor} from {url}'.format(
+                    instructor=person.full_name,
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py]: You might want uuid here? Depends on how likely you are to look them up in the DB later or you just want pretty log messages.

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                    url=person.profile_image_url)
+                )
+
+            if response and response.status_code == requests.codes.ok:
+
+                # Get the extension and check that it is in our list of image types
+                content_type = response.headers['Content-Type'].lower()
+                if content_type in IMAGE_TYPES:
+                    extension = IMAGE_TYPES.get(content_type)
+                    tmp_image_file = ContentFile(response.content, name='tmp.' + extension)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py:75]: Does this overwrite the same tmp.jpg and tmp.png each loop run? Do these files stay around on the filesystem?

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                        logger.info('Saved image for Instructor {instructor} from {url}'.format(
+                            instructor=person.full_name,
+                            url=person.profile_image_url)
+                        )
+                    else:
+                        logger.error('failed to create image file for Instructor {instructor} from {url}'.format(
+                            instructor=person.full_name,
+                            url=person.profile_image_url)
+                        )
+                        failed_image_pulls.append(person.full_name)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py]: Same uuid vs name question here.

**Review by mikix — COMMENTED** [review]
Looks reasonable. Got some questions in-line.

Also, currently this script just rips through the whole database, grabbing every image. Make sure to run this on stage before running on prod. :smile:

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py (hunk 3 above)
Comment (staubina) [reply course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py:75]: In the documentation, there is mention that files are renamed and that the original is on the filesystem with any other versions in the same location.
https://github.com/xarg/django-stdimage

**Comment by staubina** [comment]
@mikix Please take another look at this PR.

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py (hunk 3 above)
Comment (mikix) [reply course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py:75]: Another option is tempfile.TemporaryFile or tempfile.NamedTemporaryFile. Those are python-provided and can optionally delete the file when done.

ContentFile's documentation makes it sound like you're dropping and leaving files in MEDIA_ROOT, right?

File: course_discovery/apps/course_metadata/management/commands/tests/test_download_drupal_instructor_image.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import responses
+import pytest
+
+from course_discovery.apps.course_metadata.tests import factories
+from django.core.management import call_command
+from django.core.files.base import ContentFile
+
+
+@pytest.mark.django_db
+class TestDownloadDrupalInstructorImage():
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/tests/test_download_drupal_instructor_image.py:11]: This is a nit, don't worry about these if you are confident. A couple tests that are missing: multiple people (success case) & multiple people (one failed, confirm others still did the right thing).

Tests share a lot of code that could be pulled out maybe. But who cares. They're tests.

**Review by mikix — APPROVED** [review]

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+class Command(BaseCommand):
+
+    def handle(self, *args, **options):
+        config = ProfileImageDownloadConfig.get_solo()
+        person_uuids = [uuid_str.strip() for uuid_str in config.person_uuids.split(',')]
+
+        # Get the Person objects that have the marketing URL Set
+        persons = Person.objects.filter(
+            uuid__in=person_uuids
+        )
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py:26]: This works fine, but will not tell you if you entered a bad UUID for some reason (typo, bad copy/paste). Probably not worth working around, but just heads up, be careful with data entry.

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+                # Get the extension and check that it is in our list of image types
+                content_type = response.headers['Content-Type'].lower()
+                if content_type in IMAGE_TYPES:
+                    extension = IMAGE_TYPES.get(content_type)
+                    tmp_image_file = ContentFile(response.content, name='tmp.' + extension)
+                    if tmp_image_file:
+                        person.profile_image = tmp_image_file
+                        logger.info('TEST PATH -- ' + tmp_image_file.name)
+                        logger.info('TEST PATH -- ' + person.profile_image.url)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py]: Is this final code?

**Review by mikix — APPROVED** [review]

File: course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py (hunk 7 above)
Comment (staubina) [reply course_discovery/apps/course_metadata/management/commands/download_drupal_instructor_image.py]: Will remove right now.

**Pull request merged**